File: redaction.go
package main

import (
	"fmt"
	"net/url"
	"regexp"
	"strings"
)

const redactedValue = "[redacted]"
const maxUserVisibleErrorBytes = 1200

var sensitiveTextPatterns = []struct {
	pattern     *regexp.Regexp
	replacement string
}{
	{
		pattern:     regexp.MustCompile(`(?i)\b(Authorization\s*[:=]\s*)Bearer\s+([A-Za-z0-9._~+/=-]{8,})`),
		replacement: `${1}Bearer ` + redactedValue,
	},
	{
		pattern:     regexp.MustCompile(`(?i)\bBearer\s+([A-Za-z0-9._~+/=-]{8,})`),
		replacement: `Bearer ` + redactedValue,
	},
	{
		pattern:     regexp.MustCompile(`\b(sk-[A-Za-z0-9_-]{8,})\b`),
		replacement: redactedValue,
	},
	{
		pattern: regexp.MustCompile(
			`(?i)([?&](?:access_token|refresh_token|api_key|apikey|key|token|client_secret|clientsecret)=)([^&#\s]+)`,
		),
		replacement: `${1}` + redactedValue,
	},
	{
		pattern: regexp.MustCompile(
			`(?i)\b((?:api[_-]?key|apikey|access[_-]?token|refresh[_-]?token|client[_-]?secret|clientsecret|token)\s*[:=]\s*)(\[redacted\]|"[^"]*"|'[^']*'|[^\s,;)}\]]+)`,
		),
		replacement: `${1}` + redactedValue,
	},
	{
		pattern: regexp.MustCompile(
			`(?i)(["']?(?:api[_-]?key|apiKey|access[_-]?token|accessToken|refresh[_-]?token|refreshToken|client[_-]?secret|clientSecret|token)["']?\s*:\s*)(\[redacted\]|"[^"]*"|'[^']*'|[^\s,}]+)`,
		),
		replacement: `${1}` + redactedValue,
	},
}

var httpURLPattern = regexp.MustCompile(`https?://[^\s"'<>]+`)

func redactSensitiveText(value string) string {
	redacted := strings.TrimSpace(value)
	if redacted == "" {
		return ""
	}
	for _, rule := range sensitiveTextPatterns {
		redacted = rule.pattern.ReplaceAllString(redacted, rule.replacement)
	}
	if len(redacted) <= maxUserVisibleErrorBytes {
		return redacted
	}
	return redacted[:maxUserVisibleErrorBytes] + "... [truncated]"
}

func redactSensitiveValue(value any) string {
	return redactSensitiveText(fmt.Sprint(value))
}

func redactURLQueryValuesInText(value string) string {
	redacted := redactSensitiveText(value)
	if redacted == "" {
		return ""
	}
	return httpURLPattern.ReplaceAllStringFunc(redacted, func(rawURL string) string {
		parsed, err := url.Parse(rawURL)
		if err != nil || parsed.RawQuery == "" {
			return rawURL
		}
		query := parsed.Query()
		for key := range query {
			query[key] = []string{redactedValue}
		}
		parsed.RawQuery = query.Encode()
		return parsed.String()
	})
}

func redactErrorText(err error) string {
	if err == nil {
		return ""
	}
	return redactSensitiveText(err.Error())
}

File: services/pdf_service/core/redaction.py
"""Helpers for keeping credentials out of user-visible errors and logs."""

import re
from typing import Any

REDACTED = "[redacted]"
MAX_SAFE_ERROR_LENGTH = 1200

_REDACTION_PATTERNS = [
    (
        re.compile(r"\b(Authorization\s*[:=]\s*)Bearer\s+([A-Za-z0-9._~+/=-]{8,})", re.IGNORECASE),
        lambda match: f"{match.group(1)}Bearer {REDACTED}",
    ),
    (
        re.compile(r"\bBearer\s+([A-Za-z0-9._~+/=-]{8,})", re.IGNORECASE),
        f"Bearer {REDACTED}",
    ),
    (
        re.compile(r"\b(sk-[A-Za-z0-9_-]{8,})\b"),
        REDACTED,
    ),
    (
        re.compile(
            r"([?&](?:access_token|refresh_token|api_key|apikey|key|token|client_secret|clientsecret)=)([^&#\s]+)",
            re.IGNORECASE,
        ),
        lambda match: f"{match.group(1)}{REDACTED}",
    ),
    (
        re.compile(
            r"\b((?:api[_-]?key|apikey|access[_-]?token|refresh[_-]?token|client[_-]?secret|clientsecret|token)\s*[:=]\s*)(\"[^\"]*\"|'[^']*'|[^\s,;)}\]]+)",
            re.IGNORECASE,
        ),
        lambda match: f"{match.group(1)}{REDACTED}",
    ),
    (
        re.compile(
            r"([\"']?(?:api[_-]?key|apiKey|access[_-]?token|accessToken|refresh[_-]?token|refreshToken|client[_-]?secret|clientSecret|token)[\"']?\s*:\s*)(\"[^\"]*\"|'[^']*'|[^\s,}]+)",
            re.IGNORECASE,
        ),
        lambda match: f"{match.group(1)}{REDACTED}",
    ),
]


def redact_sensitive_text(value: str) -> str:
    """Redact common credential shapes from a string."""
    text = str(value or "").strip()
    if not text:
        return ""

    for pattern, replacement in _REDACTION_PATTERNS:
        text = pattern.sub(replacement, text)

    if len(text) <= MAX_SAFE_ERROR_LENGTH:
        return text
    return f"{text[:MAX_SAFE_ERROR_LENGTH]}... [truncated]"


def redact_sensitive_value(value: Any) -> Any:
    """Redact strings inside JSON-like structures."""
    if isinstance(value, str):
        return redact_sensitive_text(value)
    if isinstance(value, dict):
        return {key: redact_sensitive_value(item) for key, item in value.items()}
    if isinstance(value, list):
        return [redact_sensitive_value(item) for item in value]
    return value
